### PR TITLE
Add a prefix list real S3 e2e Rust test

### PR DESCRIPTION
**Pull request by SomeoneToIgnore**
A follow-up of https://github.com/neondatabase/neon/pull/3843#discussion_r1147757889 , that adds the first S3 integration tests

### CODE REVIEW

File: libs/remote_storage/src/s3_bucket.rs [hunk 1]
```diff
@@ -468,3 +480,33 @@ impl RemoteStorage for S3Bucket {
         Ok(())
     }
 }
+
+/// A set of utility constants and methods to use in the e2e tests.
+///
+/// `#[cfg(test)]` modules do not get shared across all things that `cargo test` launches, in particular,
+/// integrationt tests under `tests/` would not see it.
```
Comment (LizardWizzard) [inline libs/remote_storage/src/s3_bucket.rs:487]: typo
```suggestion
/// integration tests under `tests/` would not see it.
```

**Comment by SomeoneToIgnore** [comment]
Welp, the test fails with
```
2023-03-28T13:01:09.905354Z  INFO Using s3 bucket 'ci-tests-s3' in region 'us-west-2' as a remote storage, prefix in bucket: 'Some("pagination_should_work_test_1680008469905/")', bucket endpoint: 'None'

...snip 


Caused by:
    0: Error { code: "AccessDenied", message: "Access Denied", request_id: "WE44JPZ4VA7WTH6G", s3_extended_request_id: "6yYbnOwa8qBbeH3aD+qLU4P8DGcCFkd3ncRR9g/lM/C+E23Qeb5b0skZyjNSWp4rfJVVn3OrQFM=" }
    1: Error { code: "AccessDenied", message: "Access Denied", request_id: "WE44JPZ4VA7WTH6G", s3_extended_request_id: "6yYbnOwa8qBbeH3aD+qLU4P8DGcCFkd3ncRR9g/lM/C+E23Qeb5b0skZyjNSWp4rfJVVn3OrQFM=" }
    2: Error { code: "AccessDenied", message: "Access Denied", request_id: "WE44JPZ4VA7WTH6G", s3_extended_request_id: "6yYbnOwa8qBbeH3aD+qLU4P8DGcCFkd3ncRR9g/lM/C+E23Qeb5b0skZyjNSWp4rfJVVn3OrQFM=" }
```

https://github.com/neondatabase/neon/actions/runs/4543295204/jobs/8007804407#step:14:658

I know that the agents we use have an access key for the public github bucket, will try to use that.

File: libs/remote_storage/src/s3_bucket.rs [hunk 2]
```diff
@@ -337,6 +343,11 @@ impl RemoteStorage for S3Bucket {
 
         let mut document_keys = Vec::new();
 
+        #[cfg(not(test))]
+        let response_key_limit = None;
```
[the remark this answers is not in the document]
Comment (LizardWizzard) [reply libs/remote_storage/src/s3_bucket.rs:347]: This way there is no need to share the module with cosnts between impl and test code. Everything can be defined in test

**Review by LizardWizzard — COMMENTED** [review]
Overall looks pretty good, thanks!

File: libs/remote_storage/src/s3_bucket.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply libs/remote_storage/src/s3_bucket.rs]: Seems that it's indeed easy to add such field into config, this way looks way better, thanks for stopping me 🙂

**Review by SomeoneToIgnore — COMMENTED** [review]
Thank you for suggesting https://github.com/markhildreth/test-context , I've picked it over the other suggested library to get the automatic tear down implemented, that moved out quite a lot of S3 fiddling code.

File: libs/remote_storage/src/s3_bucket.rs (hunk 1 above)
Comment (SomeoneToIgnore) [reply libs/remote_storage/src/s3_bucket.rs]: No such code anymore.

**Pull request closed without merging**